src/nvim/api/private/helpers.c

```diff
@@ -1342,3 +1346,187 @@ const char *get_default_stl_hl(win_T *wp)
     return "StatusLineNC";
   }
 }
+
+void add_user_command(String name, Object command, Dict(user_command) *opts, int flags, Error *err)
+{
+  uint32_t argt = 0;
+  long def = -1;
+  cmd_addr_T addr_type_arg = ADDR_NONE;
+  int compl = EXPAND_NOTHING;
+  char *compl_arg = NULL;
+  char *rep = NULL;
+  LuaRef luaref = LUA_NOREF;
+  LuaRef compl_luaref = LUA_NOREF;
+
+  if (HAS_KEY(opts->range) && HAS_KEY(opts->count)) {
+    api_set_error(err, kErrorTypeValidation, "'range' and 'count' are mutually exclusive");
+    return;
+  }
+
+  if (opts->nargs.type == kObjectTypeInteger) {
+    switch (opts->nargs.data.integer) {
+    case 0:
+      // Default value, nothing to do
+      break;
+    case 1:
+      argt |= EX_EXTRA | EX_NOSPC | EX_NEEDARG;
+      break;
+    default:
+      api_set_error(err, kErrorTypeValidation, "Invalid value for 'nargs'");
+      return;
+    }
+  } else if (opts->nargs.type == kObjectTypeString) {
+    if (opts->nargs.data.string.size > 1) {
+      api_set_error(err, kErrorTypeValidation, "Invalid value for 'nargs'");
+    }
+
+    switch (opts->nargs.data.string.data[0]) {
+    case '*':
+      argt |= EX_EXTRA;
+      break;
+    case '?':
+      argt |= EX_EXTRA | EX_NOSPC;
+      break;
+    case '+':
+      argt |= EX_EXTRA | EX_NEEDARG;
+      break;
+    default:
+      api_set_error(err, kErrorTypeValidation, "Invalid value for 'nargs'");
+      return;
+    }
+  } else if (HAS_KEY(opts->nargs)) {
+    api_set_error(err, kErrorTypeValidation, "Invalid value for 'nargs'");
+    return;
+  }
```

Comment:
I think it would be better to switch the order of this condition by basically moving all the way to the top
```c
    if (HAS_KEY(opts->nargs)) {
      if (opts->nargs.type == kObjectTypeInteger) {
        // ...
      } else if (opts->nargs.type == kObjectTypeString) {
        // ..
      } else {
        // error
      }
    }
```
That way you won't be trying to de-reference `nargs` if it doesn't exist, and you have a secure `else` branch that acts as a proper catch-all. It's also probably easier to read as a bonus! 😃

Reply:
@muniter made the same comment [here](https://github.com/neovim/neovim/pull/16752#discussion_r774874079).

`opts->nargs` always exists. If "nargs" isn't present in the options dict, it just has type `kObjectTypeNil`. `HAS_KEY` is simply a check for `opts->nargs.type != kObjectTypeNil`. So your suggestion would be transforming the check to

```c
if (opts->nargs.type != kObjectTypeNil) {
  if (opts->nargs.type == kObjectTypeInteger)
```

but this adds an unnecessary check as well as unnecessary indentation, since the current implementation is just

```
if (opts->nargs.type == kObjectTypeInteger) {
  ...
} else if (opts->nargs.type != kObjectTypeNil) {
  // ERROR
}
```

i.e. the pattern is "check for valid types, and if the value is not nil and also not a valid type it's an error".

Reply:
I'd choose security over style in this case, since this is actually breaking a rule:

> All if…​else if constructs shall be terminated with an else statement

You could always check for errors and bail out as early as possible. Generally speaking, I think this function might be trying to do too much. I would break it into at 3-4 smaller functions with each handling a smaller an argument type. That way it's more modular and easier to swap in the future, and helps reduce the cyclomatic complexity.

Reply:
This should not be changed. Eventually keyset dicts will gain some typing support, and then refactoring code will be _much_ easier if we follow a consistent pattern instead of engaging in _ad-hoc_ reasoning in every single if-else stament for keyset arg parsing. There is no issue neither with "security" nor "style" here.

Reply:
> There is no issue neither with \"security\" nor \"style\" here.


Consistency is always a good I guess, I just figured I should point it out in case there's new precedence since it is a pretty straightforward security rule.